parser/parser.go

```diff
@@ -0,0 +1,629 @@
+	}
+
+	if r == ' ' {
+		if !skipSpaces(p) {
+			p.sendError("parser: (param state) unexpected end of input")
+			p.Consume()
+			return nil
+		}
+		return paramState
+	}
+	if r != '\x0D' { // CR
+		return trailState
+	}
+
+	return endState
+}
+
+func trailState(p *Parser) StateFn {
+	r, ok := p.Next()
+	if !ok {
+		p.sendError("parser: (trail state) unexpected end of input")
+		p.Consume()
+		return nil
+	}
+
+	if r != ':' {
+		p.sendError("parser: (trail state) expected ':' but got '" + string(r) + "'")
+		return errorState
+	}
+
+	p.Consume()
+	p.msg.Params = append(p.msg.Params, parseUntil(p, isTrailingParamRune))
+
+	r, ok = p.Next()
+	if !ok {
+		p.sendError("parser: (trail state) unexpected end of input")
+		p.Consume()
+		return nil
+	}
+
+	if r != '\x0D' { // CR
+		p.sendError("parser: (trail state) expected CR but got '" + string(r) + "'")
+		return errorState
+	}
+
+	r, ok = p.Next()
+	if !ok {
+		p.sendError("parser: (trail state) unexpected end of input")
+		p.Consume()
+		return nil
+	}
+
+	if r != '\x0A' { // LF
+		p.sendError("parser: (trail state) expected LF but got '" + string(r) + "'")
+		return errorState
+	}
+
+	return endState
+}
+
+func endState(p *Parser) StateFn {
+	p.cout <- p.msg
+	return beginState
+}
+
+func errorState(p *Parser) StateFn {
+	parseUntil(p, func(r rune) bool {
+		return r != '\r'
+	})
+	parseUntil(p, func(r rune) bool {
+		return r != '\n'
+	})
+	p.Consume()
+	return beginState
+}
+
+// parseUntil parses runes until the given predicate fails for one of the runes.
+func parseUntil(p *Parser, pred func(rune) bool) string {
+	r, ok := p.Next()
+	if !ok {
+		return ""
+	}
+	for pred(r) {
+		r, ok = p.Next()
+		if !ok {
+			return p.Consume()
+		}
+	}
+	p.Rewind()
+	return p.Consume()
+}
+
+func isHostnameRune(r rune) bool {
+	switch true {
+	case '0' <= r && r <= '9':
+	case 'A' <= r && r <= 'Z':
+	case 'a' <= r && r <= 'z':
+	case r == '.':
+	case r == '-':
+	default:
+		return false
+	}
+	return true
+}
+
+func isValueRune(r rune) bool {
+	switch r {
+	case '\x00': // NUL
+	case '\x07': // BEL
+	case '\x0D': // CR
+	case '\x0A': // LF
+	case ';':
+	case ' ':
+	default:
+		return true
+	}
+	return false
+}
+
+func isPrefixNameRune(r rune) bool {
+	switch r {
+	case '!':
+	case '@':
+	case ' ':
+	default:
+		return true
+	}
+	return false
+}
+
+func isUsernameBeginRune(r rune) bool {
+	switch {
+	case unicode.IsLetter(r):
+	case r == '[':
+	case r == ']':
+	case r == '_':
+	default:
+		return false
+	}
+	return true
+}
+
+func isUsernameRune(r rune) bool {
+	switch {
+	case unicode.IsLetter(r):
+	case unicode.IsDigit(r):
+	case r == '[':
+	case r == ']':
+	case r == '_':
+	case r == '-':
+	default:
+		return false
+	}
+	return true
+}
+
+func isCommandRuneFunc() func(r rune) bool {
+	var lastRuneCR bool
+	return func(r rune) bool {
+		switch {
+		case r == ' ':
+		case lastRuneCR && r == '\n':
+		case r == '\r':
+			lastRuneCR = true
+			return true
+		case unicode.IsLetter(r):
+			fallthrough
```

Comment:
Why does the case for when it is a letter not set `lastRuneCR`?

Reply:
I fallthrough to the case where it's set to false, i.e. the `isNumber` case. The reason it's false is because the CR case is checked as an earlier case, so we know the rune cannot be CR in any other case, such as this one.